File: 15-3sum/15-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        
        for i, v in enumerate(nums):
            if v > 0:
                continue
            if i > 0 and v == nums[i-1]:
                continue
            
            low = i+1
            high = len(nums)-1
            while (low < high):
                sum = v + nums[low] + nums[high]
                if sum == 0:
                    res.append([v, nums[low], nums[high]])
                    low += 1
                    while nums[low] == nums[low-1] and low < high:
                        low += 1
                elif sum < 0:
                    low += 1
                else:
                    high -= 1
        
        return res

File: 15-3sum/15-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        for (int i = 0; i < n-2; i++) {
            if (nums[i] > 0) {
                continue;
            }
            if (i > 0 && nums[i] == nums[i-1]) {
                continue;
            }
            int low = i+1;
            int high = n-1;
            while (low < high) {
                int sum = nums[i] + nums[low] + nums[high];
                if (sum == 0) {
                    res.push_back({nums[i], nums[low], nums[high]});
                    while (low < high && nums[low] == nums[low+1]) {
                        low++;
                    }
                    low++;
                    while (low < high && nums[high] == nums[high-1]) {
                        high--;
                    }
                    high--;
                } else if (sum < 0) {
                    low++;
                } else {
                    high--;
                }
            }
        }
        return res;
    }
};